### ISSUE TITLE
Deposit contract and address standard Discussion Thread

### ISSUE THREAD

**Issue by junderw** [junderw/deposit-contract-poc#1]
This thread is for discussions from the EIP.

**Comment by yihuang** [comment]
I think one can also do the deposit with a raw value transfer with data field set to the id. I guess the benefits of using smart contract to do that is we can refuse the transfer when system shutdown?

**Comment by junderw** [comment]
Also, you could change the account to forward to from the contract. (ie. setting some storage element to your cold account, then afterward if you change keys you can change the cold account address in storage)

Not to mention, if you send a tx to an external account with the data field for this EIP, the receiving wallet could just do what you wanted to do anyways.

So in your case, only some specialized deposit monitoring system (with tracing enabled) can manage the deposits...

But with a smart contract based approach, your case can just throw away the first 4 bytes (function signature) and the last 24 bytes (0x00) and you can still do the same thing... but now people can use smart contracts as well.

**Comment by yihuang** [comment]
I just gave it a little more thought, to unify deposit handling of native and erc20 token, would it make sense to do something similar to what you suggest [here](https://github.com/junderw/deposit-contract-poc/issues/2#issuecomment-712483712) for native token:
```solidity

contract NativeToken {
    event TransferWithId(from indexed address, to indexed address, id indexed bytes8, amount uint256);
    constructor(address payable forwardAddr) public {
        forwardAddress = forwardAddr;
    }
    function deposit(bytes8 id) external override payable returns (bool) {
        require(checksumMatch(id), 'checksum mismatch');

        (bool result, ) = forwardAddress.call{ value: msg.value }('');
        require(result, 'Could not forward funds');

        emit TransferWithId(msg.sender, forwardAddress, id, msg.value);
        return true;
    }

}
```

So one can monitor the same event type for deposits of both native and erc20 tokens.
different call signatures, same event signature, what do you think?

**Comment by junderw** [comment]
It would be an unnecessary waste of gas for each deposit.

With ERC20, there are allowances etc. so you need to log from which account the tokens were sent, as a `from` field won't really tell you where the tokens came from. You also need the `to` address to tell where the funds were sent (since all transactions for all users are made to the same smart contract, we can't assume "to" means "this contract", unlike the deposit contract)

With native value (ETH), the only reason to log msg.sender would be to say "who did these ETH coins come from?"... which the id system is supposed to be exactly that. When an exchange generates id 0x0102030405060708 for Bob... we know that if someone is using that id, then the deposit is "from Bob" in our eyes... there is no need for `from`... For `to` we know we are the `to` address always for ETH, since we know the `to` will always be `forwardAddress` there's no need to log it. We should know what our forwardAddress is and when we change it independently from the ETH blockchain.

**Comment by yihuang** [comment]
makes sense, since there's already `Transfer` event in erc20, I guess i would add something like this to existing erc20 token:
```
    event DepositTo(address indexed to, bytes8 id, uint256 value);
    function depositTo(address to, uint256 amount, bytes8 id) override public returns (void) {
        require(checksumMatch(id), 'checksum mismatch');
        require(transfer(to, amount), 'Could not forward funds');
        emit DepositTo(to, id, amount);
    }
```
it'll emit both `Transfer` and `DepositTo` events.
